File: server/src/main/java/edu/cnm/deepdive/coffeeshop/service/ShopService.java
package edu.cnm.deepdive.coffeeshop.service;

import edu.cnm.deepdive.coffeeshop.model.dto.ShopDto;
import edu.cnm.deepdive.coffeeshop.model.entity.Profile;
import edu.cnm.deepdive.coffeeshop.model.entity.Shop;
import java.util.List;
import java.util.UUID;

public interface ShopService {

  ShopDto getShop(UUID id);

  ShopDto saveShop(Shop shop, Profile profile);

  List<ShopDto> getAllShops();

}

File: app/src/main/java/edu/cnm/deepdive/coffeeshop/service/ShopService.kt
package edu.cnm.deepdive.coffeeshop.service

import edu.cnm.deepdive.coffeeshop.model.domain.Shop
import edu.cnm.deepdive.coffeeshop.repository.ShopRepository
import jakarta.inject.Inject
import jakarta.inject.Singleton
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.future.future
import java.util.UUID
import java.util.concurrent.CompletableFuture

@Singleton
class ShopService @Inject constructor(private val shopRepository: ShopRepository){

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    suspend fun getShop(id: UUID) =
        scope.future {
            shopRepository.getShop(id)
        }

    suspend fun getShops(): CompletableFuture<List<Shop>> =
        scope.future {
            shopRepository.getShops()
        }
}